velox/expression/CastHooks.h

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "velox/expression/FunctionCallToSpecialForm.h"
+#include "velox/expression/SpecialForm.h"
+#include "velox/type/Conversions.h"
+
+namespace facebook::velox::exec {
+class CastExpr;
+
+class CastHooks {
+ public:
+  virtual ~CastHooks() {}
+
+  static const std::shared_ptr<CastHooks>& get();
+
+  virtual void castTimestampToString(
```

Comment:
Sorry for the inconvenience. I improved the documentation for these hooks. For this one, I was using a "leaky" function `applyCastPrimitivesDispatch` to avoid extra if-else clause. Because legacy needs to be handled in this hook, but a virtual function cannot be template-based. I recognize it will make the API inconsistent, and hard to understand. Updated as you suggested.